File: ContigousArray.java
//Time Complexity is O(n) and space Complexity is O(n)
public class ContigousArray {
    public int findMaxLength(int[] nums) {
        if(nums.length ==0 || nums == null){
            return -1;
        }

        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0,-1);
        int rsum = 0;
        int max = 0;

        for(int i = 0; i<nums.length; i++){
            if(nums[i] == 0){
                rsum = rsum - 1;

            }else{
                rsum = rsum + 1;
            }
            if(!map.containsKey(rsum)){
                map.put(rsum, i);
            }else{
                max = Math.max(max, i - map.get(rsum));
            }

        }
        return max;
    }
}

File: ContigousArray.cpp
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// This problem is implemented using running sum approach.

class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int,int> map;
        map[0] = -1;
        int length = 0,sum = 0;

        for(int i=0;i<nums.size();i++)
        {
            sum = nums[i]?sum + 1:sum -1;
            if(map.find(sum)!=map.end())
            {
                length = max(length, i - map[sum]);
            }
            else
                map[sum] = i;
        }
        return length;


    }
};
